### PR TITLE
Ensure that DDP wrapped module has parameters that require gradients

### PR DESCRIPTION
**Pull request by danthe3rd**
…ent - see #25552

**TEST PLAN**
```
python test/run_test.py -f distributed
```

### CODE REVIEW

File: test/test_distributed.py [hunk 1]
```diff
@@ -1656,6 +1656,9 @@ def _test_DistributedDataParallel(self, gpu_subset, rank, output_device=None):
         # Run a simple end to end DDP model, use result of single node model
         # as baseline
 
+        # a module without gradients shouldn't be accepted
+        self.assertRaises(AssertionError, lambda: nn.parallel.DistributedDataParallel(nn.Module(), device_ids=gpu_subset))
+
```
Comment (pietern) [inline test/test_distributed.py]: Can you turn this into its own test case? This function is just a runner, and there is no need to re-check the assertion from every test.

Comment (danthe3rd) [reply test/test_distributed.py]: Updated thank you :)

File: test/test_distributed.py [hunk 2]
```diff
@@ -1718,6 +1718,18 @@ def test_DistributedDataParallelCPU(self):
         )
         self._barrier()
 
+    @unittest.skipIf(BACKEND != 'nccl' and BACKEND != 'gloo',
+                     "Only Nccl & Gloo backend support DistributedDataParallel")
+    @skip_if_no_cuda_distributed
+    @skip_if_no_gpu
+    def test_DistributedDataParallel_python(self):
```
Comment (pietern) [inline test/test_distributed.py]: Can you rename this to be more descriptive of what is being tested?

E.g. `test_DistributedDataParallel_requires_grad`.

File: test/test_distributed.py [hunk 3]
```diff
@@ -1718,6 +1718,18 @@ def test_DistributedDataParallelCPU(self):
...
+        group, group_id, rank = self._init_global_test()
+        rank_to_GPU = self._init_multigpu_helper()
+        gpus = list(rank_to_GPU[rank])
+
+        # a module without gradients shouldn't be accepted
+        self.assertRaises(AssertionError, lambda: nn.parallel.DistributedDataParallel(nn.Module(), device_ids=gpus))
```
Comment (pietern) [inline test/test_distributed.py]: Not strictly needed, but you can turn this into a test that doesn't need GPUs, since it doesn't allocate tensors.

**Review by pietern — APPROVED** [review]
Thanks, @danthe3rd, one more comment regarding the naming of the test.

Can you update, check if CI stays green, and then land?

**Comment by danthe3rd** [comment]
The failing tests seem unrelated to this change, as they fail with this error
```
#!/bin/bash -eo pipefail
~/workspace/.circleci/scripts/setup_linux_system_environment.sh
/bin/bash: /home/circleci/workspace/.circleci/scripts/setup_linux_system_environment.sh: No such file or directory
Exited with code 1
```

**Comment by pietern** [comment]
@danthe3rd Thanks -- I think this must be due to CI staleness. The CircleCI workspace gets clobbered after a while, so retesting no longer works on the same commit.

**Comment by danthe3rd** [comment]
So should I land it?

**Comment by pietern** [comment]
@danthe3rd Yes, let's do it! :)
